package-res/resources/web/pentaho/visual/role/mapping.js

```diff
@@ -0,0 +1,613 @@
+          // Cannot clear (monotonicity).
+          if(this === _mappingType || values == null) return;
+
+          if(values.constructor === Object) {
+            this._configureLevels(values);
+          } else if(Array.isArray(values) || (values instanceof List)) {
+            var levels = this._ensureLevelsOwn();
+
+            // TODO: Because we don't yet expose independent lists' events,
+            // we need to validate addition by hand and create a levels list to be added,
+            // thus performing parsing for us...
+            var addLevels = new ListLevel(values);
+
+            // A quantitative measurement level cannot be added if data type is qualitative.
+            var dataType = this.dataType;
+            if(!dataType.isAbstract && !isDataTypeQuantitative(dataType)) {
+              addLevels.each(function(addLevel) {
+                // New level and quantitative?
+                if(!levels.has(addLevel.key) && MeasurementLevel.type.isQuantitative(addLevel)) {
+                  throw error.argInvalid("levels",
+                      bundle.format(
+                          bundle.structured.errors.mapping.dataTypeIncompatibleWithRoleLevel,
+                          [this.dataType, addLevel]));
+                }
+              }, this);
+            }
+
+            levels.set(addLevels.toArray(), {noRemove: true, noMove: true});
+            levels.sort(MeasurementLevel.type.compare);
+          } else {
+            throw error.argInvalidType("levels", ["Array", "pentaho.type.List", "Object"], typeof values);
+          }
+        },
+
+        _ensureLevelsOwn: function() {
+          var levels = O.getOwn(this, "_levels");
+          if(!levels) {
+            // Clone the base levels list, including each element.
+            var baseLevels = this.ancestor.levels;
+            var ListType = baseLevels.constructor;
+            this._levels = levels = new ListType(baseLevels.toArray(function(elem) { return elem.clone(); }));
+          }
+          return levels;
+        },
+
+        _configureLevels: function(config) {
+          var levels = this._ensureLevelsOwn();
+          O.eachOwn(config, function(v, key) {
+            var elem = levels.get(key);
+            if(!elem) throw error.argInvalid("levels", "A level with key '" +  key + "' is not defined.");
+            elem.configure(v);
+          });
+        },
+        //endregion
+
+        //region dataType
+        _dataType: context.get(valueFactory).type,
+
+        /**
+         * Gets or sets the type of data properties required by the visual role.
+         *
+         * ### This attribute is *Monotonic*
+         *
+         * The value of a _monotonic_ attribute can change, but only in some, predetermined _monotonic_ direction.
+         *
+         * In this case, the attribute can only change to a
+         * type that is a [subtype]{@link pentaho.type.Type#isSubtypeOf} of the attribute's current value,
```

Comment:
Suggest:

type that is a [subtype]{@link pentaho.type.Type#isSubtypeOf} of the attribute's current value;